src/client.ts

```diff
@@ -284,148 +147,250 @@ export class Client {
+      // For this reason, we are going to equate the `session:started` event as "connected" which
+      // essentially means the websocket connection is stable.
+      //
+      // we are also going to let streaming client control its own connected and disconnected state so
+      // we will emit those events separately "when we are ready".
+
+      // as per block comment, we'll ignore the connected event
+      if (eventName === 'connected') {
+        return hasListeners;
+      } else if (eventName === 'disconnected') {
+        eventName = STANZA_DISCONNECTED;
+      }
+
+      return this.emit(eventName, ...args);
+    };
   }
 
-  once (eventName, ...args) {
-    if (REMAPPED_EVENTS[eventName]) {
-      (this._stanzaio.once as any)(REMAPPED_EVENTS[eventName], ...args);
-    } else {
-      (this._stanzaio.once as any)(eventName, ...args);
+  private async handleStanzaDisconnectedEvent (disconnectedInstance: NamedAgent): Promise<any> {
+    this.logger.info('stanzaDisconnected event received', { stanzaInstanceId: disconnectedInstance.id, channelId: disconnectedInstance.channelId });
+    this.connected = false;
+    this.connecting = false;
+    disconnectedInstance.pinger!.stop();
+
+    this.removeAllListeners(STANZA_DISCONNECTED);
+    this.removeAllListeners(NO_LONGER_SUBSCRIBED);
+
+    // unproxy events
+    if (disconnectedInstance.originalEmitter) {
+      disconnectedInstance.emit = disconnectedInstance.originalEmitter as any;
+    }
+
+    this.activeStanzaInstance = undefined;
+
+    this.emit('disconnected', { reconnecting: this.autoReconnect });
+
+    if (this.autoReconnect) {
+      return this.connect({ keepTryingOnFailure: true });
     }
-    return this;
   }
 
-  off (eventName, ...args) {
-    if (REMAPPED_EVENTS[eventName]) {
-      (this._stanzaio.off as any)(REMAPPED_EVENTS[eventName], ...args);
-    } else {
-      (this._stanzaio.off as any)(eventName, ...args);
+  private handleNoLongerSubscribed (stanzaInstance: NamedAgent) {
+    this.logger.warn('noLongerSubscribed event received', { stanzaInstanceId: stanzaInstance.id, channelId: stanzaInstance.channelId });
+    stanzaInstance.pinger!.stop();
+
+    this.hardReconnectRequired = true;
+
+    if (!this.reconnectOnNoLongerSubscribed) {
+      this.autoReconnect = false;
     }
-    return this;
   }
 
-  disconnect () {
+  async disconnect () {
     this.logger.info('streamingClient.disconnect was called');
-    this.stopServerLogging();
+
+    if (!this.activeStanzaInstance) {
+      return;
+    }
+
     return timeoutPromise(resolve => {
-      this._stanzaio.once('disconnected', resolve);
+      this.activeStanzaInstance!.once('disconnected', resolve);
       this.autoReconnect = false;
-      this._reconnector.stop(new Error('Cancelling reconnect')); // just in case there is already an active reconnect trying
       this.http.stopAllRetries();
-      this._stanzaio.disconnect();
+      this.activeStanzaInstance!.disconnect();
```

Comment:
I didn't know this notation existed but after a quick search, that's pretty cool and good to know!